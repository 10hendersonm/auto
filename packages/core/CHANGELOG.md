# v8.1.0 (Sat Dec 14 2019)

#### 🐛  Bug Fix

- fix bug where merging a none would skip previously meged semver label, ex: merge next into master and next has a none label  ([@hipstersmoothie](https://github.com/hipstersmoothie))

#### Authors: 1

- Andrew Lisowski ([@hipstersmoothie](https://github.com/hipstersmoothie))

---

# v8.0.0 (Wed Dec 11 2019)

#### 🐛  Bug Fix

- change skip-release releaseType to just skip  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- add hash  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- fix tests  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- fix sub-pacakge changelogs  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- only grant contributions for work in commit  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- display message when commit is omitted  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- fix bug where author would get matched incorrectly  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- handle unknown tag when supplying --from and --to  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- add tests  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- omit commits that have already been released  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- do not call afterRelease hooks during dry run  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- update docs  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- switch from dedent to endent to fix multitemplate indentation  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- add release notes to prerelease PRs  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- remove false type  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- post comment prerelase version on prerelease version PR branches  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- allow any "N > 2" of --verbose flags for very verbose  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- Deprecate "--very-verbose, -w" in favor of "-vv"  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- more logs  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- fix canary calcs  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- reset versions  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- Bump version to: v8.0.0-next.8 [skip ci]  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- change to more explicit property name  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- Bump version to: v8.0.0-next.7 [skip ci]  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- move to top level util for ease of testing  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- fix lint  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- add email fallback back in  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- add test for none  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- add none release type  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- allow user to overwrite base label  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- check for depreacted config  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- get label refactor working  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- Bump version to: v8.0.0-next.6 [skip ci]  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- move determineNextVersion to core  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- move git clean check to core  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- Bump version to: v8.0.0-next.5 [skip ci]  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- Bump version to: v8.0.0-next.4 [skip ci]  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- change getPreviousVersion hook args. can access prefixRelease from root class instead  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- Bump version to: v8.0.0-next.3 [skip ci]  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- implment dry-run flag for next command  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- Bump version to: v8.0.0-next.2 [skip ci]  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- fix test  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- shipit: add flag to only publish to 'latest' tag when "release" label is present  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- Bump version to: v8.0.0-next.1 [skip ci]  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- allow user to configure what branches are treated as prerelease branches  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- Bump version to: v8.0.0-next.0 [skip ci]  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- remove prerelease label when released  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- remove old use of prerelease label + add prerelease label to released plugin  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- Bump version to: v7.17.0-next.1 [skip ci]  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- Bump version to: v7.17.0-next.0 [skip ci]  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- must set git user before publishing so we know we can commit  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- call next from shipit on next branch  ([@hipstersmoothie](https://github.com/hipstersmoothie))
- get next branch release working  ([@hipstersmoothie](https://github.com/hipstersmoothie))

#### Authors: 1

- Andrew Lisowski ([@hipstersmoothie](https://github.com/hipstersmoothie))