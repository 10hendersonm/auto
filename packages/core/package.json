{
  "name": "@autorelease/core",
  "version": "6.5.0",
  "main": "dist/auto.js",
  "author": {
    "name": "Andrew Lisowski",
    "email": "lisowski54@gmail.com"
  },
  "bin": {
    "auto": "dist/bin/auto.js"
  },
  "publishConfig": {
    "registry": "https://registry.npmjs.org/"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/intuit/auto"
  },
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "tsc -p tsconfig.json",
    "build:watch": "npm run build -- -w",
    "lint": "tslint -p . --format stylish",
    "test": "jest --maxWorkers=2 --config ../../package.json"
  },
  "dependencies": {
    "@atomist/slack-messages": "~1.1.0",
    "@hutson/set-npm-auth-token-for-ci": "^4.0.0",
    "@octokit/graphql": "^2.1.1",
    "@octokit/plugin-enterprise-compatibility": "^1.1.0",
    "@octokit/plugin-retry": "^2.2.0",
    "@octokit/plugin-throttling": "^2.5.0",
    "@octokit/rest": "^16.24.1",
    "arr-flatten": "^1.1.0",
    "chrome-webstore-upload-cli": "^1.2.0",
    "cosmiconfig": "5.2.0",
    "dedent": "^0.7.0",
    "deepmerge": "^3.2.0",
    "dotenv": "^8.0.0",
    "enquirer": "^2.3.0",
    "env-ci": "^3.2.0",
    "get-monorepo-packages": "^1.1.0",
    "gitlog": "^3.1.2",
    "import-cwd": "^3.0.0",
    "lodash.chunk": "^4.2.0",
    "node-fetch": "2.5.0",
    "parse-author": "^2.0.0",
    "parse-commit-message": "4.0.0",
    "parse-github-url": "1.0.2",
    "registry-url": "5.1.0",
    "semver": "^6.0.0",
    "signale": "^1.4.0",
    "tapable": "^2.0.0-beta.2",
    "tinycolor2": "^1.4.1",
    "typescript-memoize": "^1.0.0-alpha.3",
    "url-join": "^4.0.0"
  },
  "devDependencies": {
    "@types/cosmiconfig": "5.0.3",
    "@types/dedent": "^0.7.0",
    "@types/dotenv": "^6.1.0",
    "@types/env-ci": "^3.1.0",
    "@types/graphql": "^14.2.0",
    "@types/jest": "~24.0.9",
    "@types/lodash.chunk": "^4.2.6",
    "@types/node": "^12.0.0",
    "@types/node-fetch": "2.3.3",
    "@types/parse-github-url": "1.0.0",
    "@types/semver": "^6.0.0",
    "@types/signale": "^1.2.1",
    "@types/tinycolor2": "^1.4.1",
    "@types/url-join": "^4.0.0",
    "all-contributors-cli": "^6.4.0",
    "graphql": "^14.2.1",
    "husky": "^2.0.0",
    "ignite": "^1.8.2",
    "jest": "~24.5.0",
    "lerna": "^3.13.4",
    "lint-staged": "^8.1.6",
    "prettier": "^1.16.4",
    "ts-jest": "^24.0.2",
    "tslint": "~5.14.0",
    "tslint-config-prettier": "~1.18.0",
    "tslint-xo": "~0.16.0",
    "typescript": "~3.3.3333",
    "typescript-tslint-plugin": "^0.3.1"
  }
}
