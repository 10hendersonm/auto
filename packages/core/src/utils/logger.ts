import signale, { Signale } from 'signale';

export interface ILogger {
  /** The level at which to log messages */
  logLevel: LogLevel;
  /** The default logger. Always on */
  log: signale.Signale<signale.DefaultMethods>;
  /** The verbose log. Has more debug logs */
  verbose: signale.Signale<signale.DefaultMethods>;
  /** The very verbose log. Has all debug logs */
  veryVerbose: signale.Signale<signale.DefaultMethods>;
}

/** Create a dummy logger for testing. */
export function dummyLog(): ILogger {
  return {
    logLevel: undefined,
    log: new Signale({ disabled: true }),
    verbose: new Signale({ disabled: true }),
    veryVerbose: new Signale({ disabled: true })
  };
}

export type LogLevel = 'verbose' | 'veryVerbose' | undefined;

/** Create a logger the the given log level. */
export default function createLog(mode: LogLevel): ILogger {
  return {
    logLevel: mode,
    log: new Signale(),
    verbose: new Signale({
      disabled: mode !== 'verbose' && mode !== 'veryVerbose'
    }),
    veryVerbose: new Signale({
      disabled: mode !== 'veryVerbose'
    })
  };
}
