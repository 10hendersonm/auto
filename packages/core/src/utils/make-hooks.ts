import {
  AsyncParallelHook,
  AsyncSeriesBailHook,
  AsyncSeriesWaterfallHook,
  SyncHook,
  SyncWaterfallHook
} from 'tapable';
import { IAutoHooks } from '../auto';
import { IChangelogHooks } from '../changelog';
import { ILogParseHooks } from '../log-parse';
import { IReleaseHooks } from '../release';

export const makeHooks = (): IAutoHooks => ({
  beforeRun: new SyncHook(['config']),
  modifyConfig: new SyncWaterfallHook(['config']),
  beforeShipIt: new SyncHook([]),
  afterShipIt: new AsyncParallelHook(['version', 'commits']),
  afterRelease: new AsyncParallelHook(['releaseInfo']),
  onCreateRelease: new SyncHook(['options']),
  onCreateChangelog: new SyncHook(['changelog', 'version']),
  onCreateLogParse: new SyncHook(['logParse']),
  getAuthor: new AsyncSeriesBailHook([]),
  getPreviousVersion: new AsyncSeriesBailHook(['prefixRelease']),
  getRepository: new AsyncSeriesBailHook([]),
  version: new AsyncParallelHook(['version']),
  afterVersion: new AsyncParallelHook([]),
  publish: new AsyncParallelHook(['version']),
  afterPublish: new AsyncParallelHook([]),
  canary: new AsyncSeriesBailHook(['canaryVersion', 'postFix'])
});

export const makeReleaseHooks = (): IReleaseHooks => ({
  onCreateChangelog: new SyncHook(['changelog', 'version']),
  createChangelogTitle: new AsyncSeriesBailHook([]),
  onCreateLogParse: new SyncHook(['logParse'])
});

export const makeLogParseHooks = (): ILogParseHooks => ({
  parseCommit: new AsyncSeriesWaterfallHook(['commit']),
  omitCommit: new AsyncSeriesBailHook(['commit'])
});

export const makeChangelogHooks = (): IChangelogHooks => ({
  renderChangelogLine: new AsyncSeriesWaterfallHook(['lineData']),
  renderChangelogTitle: new AsyncSeriesBailHook(['commits', 'lineRender']),
  renderChangelogAuthor: new AsyncSeriesBailHook([
    'author',
    'commit',
    'options'
  ]),
  renderChangelogAuthorLine: new AsyncSeriesBailHook(['author', 'user']),
  omitReleaseNotes: new AsyncSeriesBailHook(['commit'])
});
