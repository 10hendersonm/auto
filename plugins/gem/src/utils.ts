import * as fs from "fs";
import { promisify } from "util";

export const mkdir = promisify(fs.mkdir);
export const readFile = promisify(fs.readFile);
export const writeFile = promisify(fs.writeFile);
