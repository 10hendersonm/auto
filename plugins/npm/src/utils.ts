import * as fs from "fs";
import { promisify } from "util";

export const readFile = promisify(fs.readFile);
export const writeFile = promisify(fs.writeFile);

/** Check if the project is a monorepo */
export const isMonorepo = () => fs.existsSync("lerna.json");

/** Parse the lerna.json file. */
export const getLernaJson = () => {
  try {
    return JSON.parse(fs.readFileSync("lerna.json", "utf8"));
  } catch (error) {
    return {};
  }
};
