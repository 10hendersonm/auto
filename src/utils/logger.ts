import signale, { Signale } from 'signale';

export interface ILogger {
  log: signale.Signale<signale.DefaultMethods>;
  verbose: signale.Signale<signale.DefaultMethods>;
  veryVerbose: signale.Signale<signale.DefaultMethods>;
}

export function dummyLog(): ILogger {
  return {
    log: new Signale({ disabled: true }),
    verbose: new Signale({ disabled: true }),
    veryVerbose: new Signale({ disabled: true })
  };
}

export type LogLevel = 'verbose' | 'veryVerbose' | undefined;

export default function createLog(mode: LogLevel): ILogger {
  return {
    log: new Signale(),
    verbose: new Signale({
      disabled: mode !== 'verbose' && mode !== 'veryVerbose'
    }),
    veryVerbose: new Signale({
      disabled: mode !== 'veryVerbose'
    })
  };
}
