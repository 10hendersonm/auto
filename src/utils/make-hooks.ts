import { AsyncSeriesBailHook, AsyncSeriesHook, SyncHook } from 'tapable';
import { IGitHubReleaseHooks } from '../github-release';
import { IChangelogHooks } from '../log-parse';
import { IAutoHooks } from '../main';

export const makeHooks = (): IAutoHooks => ({
  beforeRun: new SyncHook(['config']),
  beforeShipIt: new SyncHook([]),
  onCreateGitHubRelease: new SyncHook(['config']),
  onCreateChangelog: new SyncHook(['gitHubReleaseConfig']),
  getAuthor: new AsyncSeriesBailHook([]),
  getPreviousVersion: new AsyncSeriesBailHook(['prefixRelease']),
  getRepository: new AsyncSeriesBailHook([]),
  publish: new AsyncSeriesHook(['version'])
});

export const makeGitHubReleaseHooks = (): IGitHubReleaseHooks => ({
  onCreateChangelog: new SyncHook(['changelog'])
});

export const makeChangelogHooks = (): IChangelogHooks => ({
  renderChangelogLine: new AsyncSeriesBailHook(['commits', 'lineRender']),
  renderChangelogTitle: new AsyncSeriesBailHook(['commits', 'lineRender']),
  renderChangelogAuthor: new AsyncSeriesBailHook([
    'author',
    'commit',
    'options'
  ]),
  renderChangelogAuthorLine: new AsyncSeriesBailHook(['author', 'user'])
});
