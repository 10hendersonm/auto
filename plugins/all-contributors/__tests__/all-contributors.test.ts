import * as Auto from "@auto-it/core";
import { execSync } from "child_process";
import { makeHooks } from "@auto-it/core/dist/utils/make-hooks";
import makeCommitFromMsg from "@auto-it/core/dist/__tests__/make-commit-from-msg";
import { dummyLog } from "@auto-it/core/dist/utils/logger";
import fs from "fs";
import env from "env-ci";

import AllContributors from "../src";

const envMock = jest.fn();
const exec = jest.fn();
const gitShow = jest.fn();
const getLernaPackages = jest.fn();

envMock.mockReturnValue({});
getLernaPackages.mockReturnValue(Promise.resolve([]));
gitShow.mockReturnValue(Promise.resolve(""));
exec.mockReturnValue("");

jest.mock("env-ci");
jest.mock("child_process");
// @ts-ignore
execSync.mockImplementation(exec);
// @ts-ignore
env.mockImplementation(envMock);
// @ts-ignore
jest.spyOn(Auto, "execPromise").mockImplementation(gitShow);
jest.spyOn(Auto, "getLernaPackages").mockImplementation(getLernaPackages);
jest.spyOn(process, "chdir").mockReturnValue();

const mockRead = (result: string) =>
  jest
    .spyOn(fs, "readFileSync")
    // @ts-ignore
    .mockReturnValueOnce(result);

describe("All Contributors Plugin", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  describe("beforeShipit", () => {
    test("should not do anything for certain release types", async () => {
      const releasedLabel = new AllContributors();
      const autoHooks = makeHooks();
      const comment = jest.fn();

      releasedLabel.apply({
        comment: comment as any,
        hooks: autoHooks,
        logger: dummyLog(),
      } as Auto.Auto);

      await autoHooks.beforeShipIt.promise({
        releaseType: "latest",
      });

      expect(comment).not.toHaveBeenCalled();

      await autoHooks.beforeShipIt.promise({
        releaseType: "old",
      });

      expect(comment).not.toHaveBeenCalled();
    });

    test("should not do anything without a PR", async () => {
      const releasedLabel = new AllContributors();
      const autoHooks = makeHooks();
      const comment = jest.fn();

      envMock.mockReturnValueOnce({ pr: undefined });

      releasedLabel.apply({
        comment: comment as any,
        hooks: autoHooks,
        logger: dummyLog(),
      } as Auto.Auto);

      await autoHooks.beforeShipIt.promise({
        releaseType: "canary",
      });

      expect(comment).not.toHaveBeenCalled();
    });

    test("should not do anything if PR doesn't exist", async () => {
      const releasedLabel = new AllContributors();
      const autoHooks = makeHooks();
      const comment = jest.fn();

      envMock.mockReturnValueOnce({ pr: 123 });

      releasedLabel.apply({
        comment: comment as any,
        hooks: autoHooks,
        logger: dummyLog(),
      } as Auto.Auto);

      await autoHooks.beforeShipIt.promise({
        releaseType: "canary",
      });

      expect(comment).not.toHaveBeenCalled();
    });

    test("should not do anything if no extra contributions found", async () => {
      const releasedLabel = new AllContributors();
      const autoHooks = makeHooks();
      const comment = jest.fn();

      envMock.mockReturnValueOnce({ pr: 123 });

      releasedLabel.apply({
        comment: comment as any,
        hooks: autoHooks,
        logger: dummyLog(),
        git: {
          getPullRequest: () => ({
            data: { body: "A body with no contributions" },
          }),
        } as any,
      } as Auto.Auto);

      await autoHooks.beforeShipIt.promise({
        releaseType: "canary",
      });

      expect(comment).not.toHaveBeenCalled();
    });

    test("should comment confirming extra contributions", async () => {
      const releasedLabel = new AllContributors();
      const autoHooks = makeHooks();
      const comment = jest.fn();

      envMock.mockReturnValueOnce({ pr: 123 });

      releasedLabel.apply({
        comment: comment as any,
        hooks: autoHooks,
        logger: dummyLog(),
        git: {
          getPullRequest: () => ({
            data: {
              body:
                "# What Changed\r\n\r\nsee title\r\n\r\n# Why\r\n\r\ncloses #1147 \r\n\r\n# Contributions\r\n\r\n- @andrew - design, doc\r\n\r\n",
            },
          }),
        } as any,
      } as Auto.Auto);

      await autoHooks.beforeShipIt.promise({
        releaseType: "canary",
      });

      expect(comment.mock.calls[0][0].message).toMatchSnapshot();
    });

    test("should warn about unknown contribution types", async () => {
      const releasedLabel = new AllContributors();
      const autoHooks = makeHooks();
      const comment = jest.fn();

      envMock.mockReturnValueOnce({ pr: 123 });

      releasedLabel.apply({
        comment: comment as any,
        hooks: autoHooks,
        logger: dummyLog(),
        git: {
          getPullRequest: () => ({
            data: {
              body:
                "# What Changed\r\n\r\nsee title\r\n\r\n# Why\r\n\r\ncloses #1147 \r\n\r\n# Contributions\r\n\r\n- @andrew - design123, doc\r\n\r\n",
            },
          }),
        } as any,
      } as Auto.Auto);

      await autoHooks.beforeShipIt.promise({
        releaseType: "canary",
      });

      expect(comment.mock.calls[0][0].message).toMatchSnapshot();
    });
  });

  test("should do nothing for username-less commits", async () => {
    const releasedLabel = new AllContributors();
    const autoHooks = makeHooks();
    mockRead('{ "contributors": [] }');

    releasedLabel.apply({ hooks: autoHooks, logger: dummyLog() } as Auto.Auto);

    await autoHooks.afterAddToChangelog.promise({
      bump: Auto.SEMVER.patch,
      currentVersion: "0.0.0",
      lastRelease: "0.0.0",
      releaseNotes: "",
      commits: [
        makeCommitFromMsg("Do the stuff"),
        makeCommitFromMsg("Do the thing", {
          files: ["src/index.ts"],
        }),
      ],
    });

    expect(exec).not.toHaveBeenCalled();
  });

  test("should find a single contribution", async () => {
    const releasedLabel = new AllContributors();
    const autoHooks = makeHooks();
    mockRead('{ "contributors": [] }');

    releasedLabel.apply({ hooks: autoHooks, logger: dummyLog() } as Auto.Auto);

    await autoHooks.afterAddToChangelog.promise({
      bump: Auto.SEMVER.patch,
      currentVersion: "0.0.0",
      lastRelease: "0.0.0",
      releaseNotes: "",
      commits: [
        {
          subject: "Do the thing",
          hash: "123",
          labels: [],
          files: ["src/index.ts"],
          authors: [{ username: "Jeff", hash: "123" }],
        },
      ],
    });

    expect(exec.mock.calls[0][0]).toBe(
      "npx all-contributors-cli add Jeff code"
    );
  });

  test("should work for single package", async () => {
    const releasedLabel = new AllContributors();
    const autoHooks = makeHooks();

    mockRead('{ "contributors": [] }');
    getLernaPackages.mockRejectedValueOnce(new Error("test"));

    releasedLabel.apply({ hooks: autoHooks, logger: dummyLog() } as Auto.Auto);

    await autoHooks.afterAddToChangelog.promise({
      bump: Auto.SEMVER.patch,
      currentVersion: "0.0.0",
      lastRelease: "0.0.0",
      releaseNotes: "",
      commits: [
        {
          subject: "Do the thing",
          hash: "123",
          labels: [],
          files: ["src/index.ts"],
          authors: [{ username: "Jeff", hash: "123" }],
        },
      ],
    });

    expect(exec.mock.calls[0][0]).toBe(
      "npx all-contributors-cli add Jeff code"
    );
  });

  test("should find contributions from merge commit", async () => {
    const releasedLabel = new AllContributors();
    const autoHooks = makeHooks();
    mockRead('{ "contributors": [] }');
    gitShow.mockReturnValueOnce("src/index.ts");

    releasedLabel.apply({ hooks: autoHooks, logger: dummyLog() } as Auto.Auto);

    await autoHooks.afterAddToChangelog.promise({
      bump: Auto.SEMVER.patch,
      currentVersion: "0.0.0",
      lastRelease: "0.0.0",
      releaseNotes: "",
      commits: [
        {
          subject: "Do the thing",
          hash: "123",
          labels: [],
          files: ["src/index.ts"],
          authors: [{ username: "Jeff", hash: "123" }],
        },
      ],
    });

    expect(exec.mock.calls[0][0]).toBe(
      "npx all-contributors-cli add Jeff code"
    );
  });

  test("should find a multiple contributions", async () => {
    const releasedLabel = new AllContributors();
    const autoHooks = makeHooks();
    mockRead('{ "contributors": [] }');

    releasedLabel.apply({ hooks: autoHooks, logger: dummyLog() } as Auto.Auto);

    await autoHooks.afterAddToChangelog.promise({
      bump: Auto.SEMVER.patch,
      currentVersion: "0.0.0",
      lastRelease: "0.0.0",
      releaseNotes: "",
      commits: [
        {
          subject: "Do the thing",
          hash: "123",
          labels: [],
          files: ["src/index.ts"],

          authors: [{ username: "Jeff", hash: "123" }],
        },
        {
          subject: "Do other thing",
          hash: "123",
          labels: [],
          files: ["src/index.test.ts"],
          authors: [{ username: "Jeff", hash: "123" }],
        },
      ],
    });

    expect(exec.mock.calls[0][0]).toBe(
      "npx all-contributors-cli add Jeff code,test"
    );
  });

  test("should update sub-packages", async () => {
    const releasedLabel = new AllContributors();
    const autoHooks = makeHooks();

    getLernaPackages.mockReturnValueOnce(
      Promise.resolve([
        {
          path: "packages/app",
          name: "@foobar/app",
          version: "1.2.3",
        },
        {
          path: "packages/lib",
          name: "@foobar/lib",
          version: "1.2.3",
        },
      ])
    );

    releasedLabel.apply({ hooks: autoHooks, logger: dummyLog() } as Auto.Auto);

    await autoHooks.afterAddToChangelog.promise({
      bump: Auto.SEMVER.patch,
      currentVersion: "0.0.0",
      lastRelease: "0.0.0",
      releaseNotes: "",
      commits: [
        {
          subject: "Do the thing",
          hash: "123",
          labels: [],
          files: ["packages/app/src/index.ts"],
          authors: [{ username: "Jeff", hash: "123" }],
        },
        {
          subject: "Do the docs",
          hash: "124",
          labels: [],
          files: ["packages/lib/README.md"],
          authors: [{ username: "Jeff", hash: "124" }],
        },
      ],
    });

    // root contributors
    expect(exec.mock.calls[0][0]).toBe(
      "npx all-contributors-cli add Jeff code,doc"
    );
    // app contributors
    expect(exec.mock.calls[1][0]).toBe(
      "npx all-contributors-cli add Jeff code"
    );
    // lib contributors
    expect(exec.mock.calls[2][0]).toBe("npx all-contributors-cli add Jeff doc");
  });

  test("should update old contributors", async () => {
    const releasedLabel = new AllContributors();
    const autoHooks = makeHooks();
    mockRead(
      '{ "contributors": [ { "login": "Jeff", "contributions": ["infra"] } ] }'
    );

    releasedLabel.apply({ hooks: autoHooks, logger: dummyLog() } as Auto.Auto);

    await autoHooks.afterAddToChangelog.promise({
      bump: Auto.SEMVER.patch,
      currentVersion: "0.0.0",
      lastRelease: "0.0.0",
      releaseNotes: "",
      commits: [
        {
          subject: "Do the thing",
          hash: "123",
          labels: [],
          files: ["src/index.ts"],
          authors: [{ username: "Jeff", hash: "123" }],
        },
      ],
    });

    expect(exec.mock.calls[0][0]).toBe(
      "npx all-contributors-cli add Jeff infra,code"
    );
  });

  test("should include sub-commit contributors", async () => {
    const releasedLabel = new AllContributors();
    const autoHooks = makeHooks();
    mockRead(
      '{ "contributors": [ { "login": "Jeff", "contributions": ["infra"] } ] }'
    );

    releasedLabel.apply({ hooks: autoHooks, logger: dummyLog() } as Auto.Auto);

    await autoHooks.afterAddToChangelog.promise({
      bump: Auto.SEMVER.patch,
      currentVersion: "0.0.0",
      lastRelease: "0.0.0",
      releaseNotes: "",
      commits: [
        {
          subject: "Do the thing",
          hash: "123",
          labels: [],
          files: ["src/index.ts"],
          authors: [
            { username: "Jeff", hash: "123" },
            { username: "Adam", hash: "456" },
          ],
        },
      ],
    });

    expect(exec.mock.calls[0][0]).toBe(
      "npx all-contributors-cli add Jeff infra,code"
    );
    expect(exec.mock.calls[1][0]).toBe(
      "npx all-contributors-cli add Adam code"
    );
  });

  test("should not update if no new contribution types", async () => {
    const releasedLabel = new AllContributors();
    const autoHooks = makeHooks();
    mockRead(
      '{ "contributors": [ { "login": "Jeff", "contributions": ["code"] } ] }'
    );

    releasedLabel.apply({ hooks: autoHooks, logger: dummyLog() } as Auto.Auto);

    await autoHooks.afterAddToChangelog.promise({
      bump: Auto.SEMVER.patch,
      currentVersion: "0.0.0",
      lastRelease: "0.0.0",
      releaseNotes: "",
      commits: [
        makeCommitFromMsg("Do the thing", {
          files: ["src/index.ts"],
          username: "Jeff",
        }),
      ],
    });

    expect(exec).not.toHaveBeenCalled();
  });
});
