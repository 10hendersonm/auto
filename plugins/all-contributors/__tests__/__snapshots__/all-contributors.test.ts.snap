// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`All Contributors Plugin beforeShipit should comment confirming extra contributions 1`] = `
"# Extra Contributions

The following contributions will be added to all-contributors (as well as any code contributions) when this PR is released :tada::

- @andrew - design, doc"
`;

exports[`All Contributors Plugin beforeShipit should warn about unknown contribution types - no valid 1`] = `
"# Extra Contributions

No valid contribution types found!

## Unknown Contribution Types

We found some unknown contribution types in your PR body!
These contributions will not be counted and you should fix them.

- \`design123\`,- \`doc456\`"
`;

exports[`All Contributors Plugin beforeShipit should warn about unknown contribution types - some valid 1`] = `
"# Extra Contributions

The following contributions will be added to all-contributors (as well as any code contributions) when this PR is released :tada::

- @adam - doc

## Unknown Contribution Types

We found some unknown contribution types in your PR body!
These contributions will not be counted and you should fix them.

- \`design123\`,- \`doc456\`"
`;

exports[`All Contributors Plugin beforeShipit should warn about unknown contribution types 1`] = `
"# Extra Contributions

The following contributions will be added to all-contributors (as well as any code contributions) when this PR is released :tada::

- @andrew - doc

## Unknown Contribution Types

We found some unknown contribution types in your PR body!
These contributions will not be counted and you should fix them.

- \`design123\`"
`;
